Optimized bandwidth and alpha for KDE with cross validation on data

jam-sadiq opened the pull request:
adding a new class for optimized bandwidth and alpha with cross validation based on log-likelihood value on training data

jam-sadiq committed:
fix bugs of incorrect names, added some code

jam-sadiq committed:
added some basic code

jam-sadiq committed:
add class for choosing bwds and alphs from cross validation based on log likelihood

jam-sadiq committed:
soem addition correction

jam-sadiq committed:
fixed some code

jam-sadiq committed:
fix close of plot

jam-sadiq committed:
fixe some issues

jam-sadiq committed:
removed extra print statements

jam-sadiq committed:
remove redundant coded line

jam-sadiq committed:
Merge branch 'main' of github.com:tdent/pop-de into opt_parameter_kde

jam-sadiq committed:
Merge branch 'tdent:main' into opt_parameter_kde

jam-sadiq committed:
Merge branch 'opt_parameter_kde' of github.com:JAMSADIQ/pop-de into opt_parameter_kde

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+        return best_params, best_score
+
+
+
```
```suggestion

```

JAMSADIQ replied:
removed

tdent commented on popde/adaptive_kde.py:
```diff
@@ -2,6 +2,8 @@
 import scipy
 from density_estimate import VariableBwKDEPy 
 from scipy.stats import gmean
+from sklearn.model_selection import KFold, LeaveOneOut
```
it is probably better to import these only when needed, ie inside the relevant methods, since it would be good to be able to run the code for other classes (ie if using a class without grid optimization) without needing to install and import sklearn

JAMSADIQ replied:
fixed and called in relevant method

tdent commented on popde/adaptive_kde.py:
```diff
@@ -2,6 +2,8 @@
 from density_estimate import VariableBwKDEPy 
 from scipy.stats import gmean
+from sklearn.model_selection import KFold, LeaveOneOut
+import matplotlib.pyplot as plt
```
similarly, only import when needed

JAMSADIQ replied:
done

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
 
+class KDEOptimization(AdaptiveBwKDE):
+    """
+    Given grid of bandwidth and alpha values 
+    find optimized values using cross 
+    valitation with figure of merit
+    based on loglikelihood criteria 
```
```suggestion
    Optimize bandwidth and alpha by grid search using
    cross validation with a log likelihood figure of merit 
```

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+    valitation with figure of merit
+    based on loglikelihood criteria 
+    """
+    def __init__(self, data, weights, bandwidth_options, alpha_options , input_transf=None, stdize=False, rescale=None, backend='KDEpy', bandwidth=0.5, alpha=0.0, dim_names=None, do_fit=True):
```
I think weights should be a kwarg with default None. 
Also why is bandwidth being given a default value 0.5?  I guess we don't need any default value as this is ignored but 0.5 seems a weird choice ..

JAMSADIQ replied:
added as kwarg

JAMSADIQ replied:
bw value is 1 by default

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+
+    def loo_cv_score(self, bandwidth_val, alpha_val):
+        loo = LeaveOneOut() #sklearn way
+        weights = self.weights
```
this line doesn't seem to do anything ?

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+           
+            #######There may need some hints how to use it better 
+            ### also if we need to add weight option here
+            local_weights = None
```
I don't quite understand this comment .. it seems what you are doing is to optimize without using any weights at present, correct ? 
If so, make a comment that explains this and also a 'FIXME' comment to suggest that we might want to use the KDE data weights in future

JAMSADIQ replied:
need more code to fix

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+            ### also if we need to add weight option here
+            local_weights = None
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
+            fom += awkde.evaluate(test_data)  
```
fom was not defined earlier, is it a float or a list or what?  .. also does this actually calculate log likelihood?

JAMSADIQ replied:
fom  initialized as a float and used log likelihood.

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+        return sum(fom)
+
+    def optimize_parameters(self, method='loo_cv', fom_plot=False):
+        best_score = float('inf')
```
I think numpy inf would be preferable ..

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+        for bandwidth in self.bandwidth_options:
+            for alpha in self.alpha_options:
+                if method=='kfold_cv':
+                    score = self.kfold_cv_score(bandwidth, alpha, n_splits=2)
```
need n_splits to be controllable via an init option

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+                    best_params['alpha'] = alpha
+        
+        if fom_plot==True:
+            import operator
```
since operator is a standard python library, I think it's OK to import it at the start of the module

JAMSADIQ replied:
done

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+            for bw in self.bandwidth_options:
+                FOMlist = [FOM[(bw, al)] for al in self.alpha_options]
+                if bw not in ['silverman', 'scott']:
+                    bw = float(bw) #bwchoice.astype(np.float) #for list
+                    ax.plot(self.alpha_options, FOMlist, label='{0:.3f}'.format(bw))
```
```suggestion
                    ax.plot(self.alpha_options, FOMlist, label='{0:.3f}'.format(float(bw)))
```

JAMSADIQ replied:
fix, may  need a  review

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+                    ax.plot(self.alpha_options, FOMlist, label='{0:.3f}'.format(bw))
+                else:
+                    ax.plot(alphagrid, FOMlist, label='{}'.format(bw))
+            if optbw not in ['silverman', 'scott']:
```
I don't think you need two separate 'if optbw' blocks (here and line 188), can you combine them?

JAMSADIQ replied:
done. KDEpy cannot take scott and silvermann option for multidimensional KDE. need some suggestion here

tdent replied:
I am not sure what you need here. We do not need to support 'scott' and 'silverman' in general so I don't know what the problem is

JAMSADIQ replied:
done

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+            handles, labels = ax.get_legend_handles_labels()
+            lgd = ax.legend(handles, labels, loc='upper center', bbox_to_anchor=(0.5,1.25), ncol =6, fancybox=True, shadow=True, fontsize=8)
+        #plt.ylim(maxFOM -5 , maxFOM +6)
+            plt.savefig("FOMfortwoDsourcecase.png", bbox_extra_artists=(lgd, ), bbox_inches='tight')
```
can you explain what this `bbox_extra_artists` is? 
Also, it's not a good idea to always save the plot with the same name, see my suggestion to allow different names to be chosen when calling the method.

JAMSADIQ replied:
just to fix the legends. added a comment, a suggestion may be better

jam-sadiq committed:
remove extra spacing and fix FOM estimate

jam-sadiq committed:
import sklearn inside method

jam-sadiq committed:
import inside method and remove extra lines

jam-sadiq committed:
fixed comment

jam-sadiq committed:
several changes based on comments

JAMSADIQ commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            local_weights = None
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
+            fom.append(awkde.evaluate(test_data).sum())
```
now it does, need review if this is correct

jam-sadiq committed:
minor changes

JAMSADIQ commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+                    score = self.loo_cv_score(bandwidth, alpha)
+
+                FOM[(bandwidth, alpha)] = score
+                if score < best_score:
```
corrected with modification. review will be good

jam-sadiq committed:
split a line

jam-sadiq committed:
fix  a small bug

JAMSADIQ commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +118,98 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+            fom.append(awkde.evaluate(test_data).sum())
+        return sum(fom)
+
+    def optimize_parameters(self, method='loo_cv', fom_plot=False):
```
fix, need review

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +117,100 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+
+    def loo_cv_score(self, bandwidth_val, alpha_val):
+        from sklearn.model_selection import LeaveOneOut
+        loo = LeaveOneOut() #sklearn way
```
```suggestion
        loo = LeaveOneOut()
```

JAMSADIQ replied:
done

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +117,100 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+        fom = 0.0
+        for train_index, test_index in loo.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            #for weights we need corressponding weights as well
```
I don't understand this comment ...

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +117,100 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+        for train_index, test_index in loo.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            #for weights we need corressponding weights as well
+            if self.weights is not None:
+                local_train_weights = weights[train_index]
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
```
I am not seeing what the line with local_train_weights is for (also in the kfold case). If the code does nothing at the moment then it is best not to write code at all but instead leave comments discussing possible improvements

JAMSADIQ replied:
preparing kde with weights  need weight at the training points.  maybe important or disregard for optimized parameter, in progress

tdent replied:
but are you saying that you are going to include something to do this in this PR, or not? 
I am saying we should not have 'in progress' code that does not do anything in the repo.
so can we please remove the lines about `local_train_weights` and make notes for any relevant development in another issue or another PR?

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +117,100 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
+            fom += np.log(awkde.evaluate(test_data))
+        return np.mean(fom)
```
do we need mean here?  It looks like fom is just a float.

JAMSADIQ replied:
corrected

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +117,100 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+        optbw, optalpha  = optval[0], optval[1]
+        best_score = FOM[(optbw, optalpha)]
+        #set self.bandwidth  and self.alpha  as optimized values
+        self.bandwidth  = optbw
```
do we not want to use `set_adaptive_parameter` here (method of AdaptiveKDE that both sets the parameters and refits the KDE automatically ?)

JAMSADIQ replied:
correct. done

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +117,100 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+    Optimize bandwidth and alpha by grid search using
+    cross validation with a log likelihood figure of merit 
+    """
+    def __init__(self, data, bandwidth_options, alpha_options, weights=None, input_transf=None, stdize=False, rescale=None, backend='KDEpy', bandwidth=1.0, alpha=0.0, dim_names=None, do_fit=True, n_splits=2):
```
I don't think we want `do_fit=True` here, since this would set up the KDE with bw=1 and alpha=0.  Ie we want to initialize all the KDE data but not actually train/fit the KDE until the optimization step.

JAMSADIQ replied:
removed here in super_init_. need a review

tdent replied:
I meant that we should have do_fit=False ... have fixed it in an edit myself

JAMSADIQ replied:
ok

jam-sadiq committed:
several changes based on comments

tdent committed:
set do_fit=False, other minor changes

jam-sadiq committed:
change FOM complicated code

jam-sadiq committed:
fix with origin

jam-sadiq committed:
remove lines, in fom plot for bw vals, and for weights

jam-sadiq committed:
minor fix

tdent commented on popde/adaptive_kde.py:
```diff
@@ -117,6 +116,88 @@ def set_adaptive_parameter(self, new_alpha, new_global_bw):
+        for train_index, test_index in loo.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, bandwidth=bandwidth_val,alpha=alpha_val)
```
so here, although we have `input_transf=None, stdize=False, rescale=None` as defaults, it may be that the user initializes the class with some input transformations, and then we will want to apply these to the optimization as well (otherwise we will not get suitable bw/alpha values).  So please also pass on the values of `self.input_transf`, `self.stdize` and `self.rescale` here

JAMSADIQ replied:
so I am bit confuse here. so I can add these for the data but when we evaluate kde on testing data (for score) do we need to also use evaluate_with_transf.

tdent replied:
Can you try to work it out for yourself please ...
We have the original data and we have the transformed data used for KDE fitting
So when doing the cross validation there are some choices that make sense and some that do not
Make one choice that makes sense and code it and I will review the code

JAMSADIQ replied:
I did make changes and tested the code.

tdent reviewed (changes requested):
This is basically ready, except for the point about data transformations which need to be passed on to the optimization / FOM calculation.

jam-sadiq committed:
added transform of data options

tdent commented on popde/adaptive_kde.py:
I don't see why you have added all these lines here, aren't they already done inside the init method of the parent class ie super.__init__ ?

tdent replied:
I mean lines 130 - 136

jam-sadiq committed:
fix use of self.kde_data in optimization and removed extra line

tdent committed:
minor fixes to optimized awkde

- Pass do_fit to parent class init
- Don't use evaluate with transf for kde data
- minor formatting fixes

tdent committed:
Actually remove module level matplotlib import

tdent committed:
small change to class name

More specific to optimization of adaptive params

tdent reviewed (approved):
I think all the issues should have been resolved, if you think it looks OK then please merge

jam-sadiq committed:
fix indent of a line

JAMSADIQ commented:
While testing the code on a twoD data with 
import adaptive_kde as adp_kde
kde = adp_kde.AdaptiveKDEOptimization(data, bwgrid, alpgrid)
I get an error:
    _```
In adaptive_kde.py", line 130, in __init__
    super().__init__(data, weights, input_transf, stdize, rescale, backend,
    adaptive_kde.py", line 47, in __init__
    self.pilot_values = self.evaluate(self.kde_data)
    density_estimate.py", line 370, in evaluate_KDEpy
    density_values = self.kernel_estimate.evaluate(points)
    AttributeError: 'NoneType' object has no attribute 'evaluate
```_

tdent commented:
OK, so we need to allow `AdaptiveBwKDE` to either have do_fit=True or do_fit=False options.  The problem is that the False option is allowed during super().__init__ but then it is assumed that the kde fit exists.  In any case this class should not produce an error if the option is False.

jam-sadiq committed:
fetch changes for fixing do_fit issue

tdent committed:
add back blank line

jam-sadiq committed:
fix evaluate_with_transf method to remove dependency of stdize and rescale ont input_transf default choice

jam-sadiq committed:
Merge branch 'opt_parameter_kde' of github.com:JAMSADIQ/pop-de into opt_parameter_kde

JAMSADIQ commented:
Now, with minor fix in density estimate module, this seems to be ready to merge

tdent commented on popde/density_estimate.py:
```diff
@@ -172,16 +172,18 @@ def evaluate_with_transf(self, points):
-            elif option in ('none', 'None'):
-                input_Jacobian = 1. 
+        for i in range(self.ndim):
+            if self.input_transf is not None: 
```
this should be before the 'for' loop, no?

jam-sadiq committed:
fix Jacobians

jam-sadiq committed:
fix the rescaling factor

jam-sadiq committed:
reuse enumerate, but with default choice Jacbian for stdize and rescaling will not be computed

jam-sadiq committed:
fix the code

tdent committed:
Simplify Jacobian calc and move evaluate transf method

tdent reviewed (approved):
Check if the slight rewrite of the stdize / rescaling Jacobians looks OK, if so then you can go ahead to squash and merge
